system-parachains/asset-hubs/asset-hub-kusama/src/impls.rs

```diff
@@ -0,0 +1,309 @@
+					let _ = T::Assets::resolve(who, credit_in).defensive();
+					return Err(InvalidTransaction::Payment.into())
+				},
+			};
+
+			ensure!(change.peek() == Zero::zero(), InvalidTransaction::Payment);
+
+			Ok((Some(fee_credit), fee, asset_fee))
+		}
+		fn correct_and_deposit_fee(
+			who: &<T>::AccountId,
+			dispatch_info: &DispatchInfoOf<<T>::RuntimeCall>,
+			post_info: &PostDispatchInfoOf<<T>::RuntimeCall>,
+			corrected_fee: Self::Balance,
+			tip: Self::Balance,
+			fee_paid: LiquidityInfoOf<T>,
+			_received_exchanged: Self::LiquidityInfo,
+			asset_id: Self::AssetId,
+			initial_asset_consumed: T::Balance,
+		) -> Result<T::Balance, TransactionValidityError> {
+			let fee_paid = if let Some(fee_paid) = fee_paid {
+				fee_paid
+			} else {
+				return Ok(Zero::zero());
+			};
+			// Try to refund if the fee paid is more than the corrected fee and the account was not
+			// removed by the dispatched function.
+			let (fee, fee_in_asset) = if fee_paid.peek() > corrected_fee &&
+				T::Assets::total_balance(asset_id.clone(), who) > Zero::zero()
+			{
+				let refund_amount = fee_paid.peek().saturating_sub(corrected_fee);
+				// Check if the refund amount can be swapped back into the asset used by `who` for
+				// fee payment.
+				let refund_asset_amount =
+					AssetConversion::<T>::quote_price_exact_tokens_for_tokens(
+						A::get(),
+						asset_id.clone(),
+						refund_amount,
+						true,
+					)
+					// No refund given if it cannot be swapped back.
+					.unwrap_or(Zero::zero());
+
+				// Deposit the refund before the swap to ensure it can be processed.
+				let debt = match T::Assets::deposit(
+					asset_id.clone(),
+					&who,
+					refund_asset_amount,
+					Precision::BestEffort,
+				) {
+					Ok(debt) => debt,
+					// No refund given since it cannot be deposited.
+					Err(_) => fungibles::Debt::<T::AccountId, T::Assets>::zero(asset_id.clone()),
+				};
+
+				if debt.peek() == Zero::zero() {
+					// No refund given.
+					(fee_paid, initial_asset_consumed)
+				} else {
+					let (refund, fee_paid) = fee_paid.split(refund_amount);
+					match S::swap_exact_tokens_for_tokens(
+						vec![A::get(), asset_id],
+						refund,
+						Some(refund_asset_amount),
+					) {
+						Ok(refund_asset) => {
+							match refund_asset.offset(debt) {
+								Ok(SameOrOther::None) => {},
+								// This arm should never be reached, as the  amount of `debt` is
+								// expected to be exactly equal to the amount of `refund_asset`
+								// credit.
+								_ => return Err(InvalidTransaction::Payment.into()),
+							};
+							(
+								fee_paid,
+								initial_asset_consumed.saturating_sub(refund_asset_amount.into()),
+							)
+						},
+						// The error should not occur since swap was quoted before.
+						Err((refund, _)) => {
+							match T::Assets::settle(who, debt, Preservation::Expendable) {
+								Ok(dust) => ensure!(
+									dust.peek() == Zero::zero(),
+									InvalidTransaction::Payment
```

Comment:
Debugging this will be fund, as it always returns `InvalidTransaction::Payment` 🙃  maybe some logs?